### DISCUSSION TITLE
Error with set-env on self hosted runner ==> [error]No such file or directory

### DISCUSSION THREAD

**Discussion by rboekweg** [community/community#26420] [Actions]
<p>While testing use of “set-env” on a self-hosted runner (2.164.0), I get “No such file or directory” error.  </p>
<p>Run echo ::set-env name=ENVIRONMENT::staging<br>
echo ::set-env name=ENVIRONMENT::staging<br>
shell: /usr/bin/bash -e {0}<br>
##[error]No such file or directory</p>
<p>I’ve tried a few variations, but all the same result. I’m not able to get any of the commands mentioned on <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/development-tools-for-github-actions#set-an-environment-variable-set-env" rel="nofollow noopener">this page</a> to work.  Anyone else running into these problems? I see a couple of new versions of the <a href="https://github.com/actions/runner/releases" rel="nofollow noopener">runner</a>, but not officially released yet.</p>

**Comment by ethomson** [comment]
<p>Is your repository public?  If so, can you point me to the workflow?  If not, can you copy and paste your workflow file?</p>

**Comment by rboekweg** [comment]
<p>Not a public repo, but this is relevant and obfuscated workflow file. The clean workspace step works fine, but the “set-env” step fails every time with error, “[error]No such file or directory”</p>
<pre><code>name: Build and Deploy ProjectX

on:
  push:
    branches:
      - master
      - xxxxxx
      - xxxxxx
jobs:
  ProjectX_build_job:
    runs-on: self-hosted
    name: Build ProjectX
    steps:
      - name: Clean Workspace
        run: sudo rm -rf $GITHUB_WORKSPACE
      - name: Setup for MASTER branch
        if: endsWith(github.ref, 'master')
        run: |
          echo ::set-env name=ENVIRONMENT::staging
          echo ::set-env name=SLACKWEBHOOK::${{ secrets.STAGING_SLACK_WEBHOOK_URL }}
...
</code></pre>

**Comment by rboekweg** [comment]
<p>I replied, but the response got marked as spam and removed?</p>

**Comment by rboekweg** [comment]
<p>Trying to upload snippet of workflow YAML one more time…</p>
<pre><code>name: Build and Deploy ProjectX

on:
  push:
    branches:
      - master
      - xxxx
      - yyyy

jobs:

  projectx_build_job:
    runs-on: self-hosted
    name: Build ProjectX
    steps:
      - name: Clean workspace
        run: sudo rm -rf $GITHUB_WORKSPACE

      - name: Setup for MASTER branch
        if: endsWith(github.ref, 'master')
        run: |
          echo ::set-env name=ENVIRONMENT::staging
          echo ::set-env name=SLACKWEBHOOK::${{ secrets.STAGING_SLACK_WEBHOOK_URL }}
...
</code></pre>

**Comment by ethomson** [comment]
<p>This looks like an issue with your self-hosted runner - in particular, it looks like you can’t find the echo command.  This seems like it may be a problem with your PATH?  What if you try <code>/bin/echo</code> instead of <code>echo</code>?</p>

**Comment by rboekweg** [comment]
<p>I suspected something like that at first, but echo and /bin/echo appear to work.  I’ve tested from an action and also when SSH session.  Both are running as default ec2-user.  </p>

**Comment by rboekweg** [comment]
<p>After turning on debugging for steps/runners (see <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/managing-a-workflow-run#enabling-debug-logging" rel="nofollow noopener">enabling debug logging</a>) I’m seeing more details.  Doing more digging into those details.</p>

**Comment by rboekweg** [comment]
<p>From debug output, the first command with echo works just fine:</p>
<pre><code>2020-02-05T17:10:57.0357003Z ##[group]Run sudo rm -rf $GITHUB_WORKSPACE
2020-02-05T17:10:57.0357730Z [36;1msudo rm -rf $GITHUB_WORKSPACE[0m
2020-02-05T17:10:57.0369379Z shell: /usr/bin/bash -e {0}
2020-02-05T17:10:57.0369945Z ##[endgroup]
2020-02-05T17:10:57.0387553Z ##[debug]/usr/bin/bash -e /mnt2/actions-runner/_work/_temp/af6bc143-becf-460a-8d01-bf7909199eb3.sh
2020-02-05T17:10:57.0639965Z ##[debug]Finishing: Clean workspace
</code></pre>
<p>But the command with the set-env fails.</p>
<pre><code>2020-02-05T17:10:57.0745992Z ##[group]Run echo ::set-env name=action_state::yellow
2020-02-05T17:10:57.0746488Z [36;1mecho ::set-env name=action_state::yellow[0m
2020-02-05T17:10:57.0753669Z shell: /usr/bin/bash -e {0}
2020-02-05T17:10:57.0754045Z ##[endgroup]
2020-02-05T17:10:57.0764310Z ##[debug]/usr/bin/bash -e /mnt2/actions-runner/_work/_temp/49b66d49-13f9-47b3-8283-9870cf2dc0f0.sh
2020-02-05T17:10:57.0795015Z ##[error]No such file or directory
</code></pre>
<p>Is there any way to preserve the temp .sh files.  Real curious to see inside these.  They get cleaned up after build finishes.</p>

**Comment by rboekweg** [comment]
<p>I copied the files before they were cleaned up.  Nothing more than the actual commands.</p>

**Comment by rboekweg** [comment, marked as answer]
<p>I’ve found the answer after more testing.  Looks like something fishy with this step that was running before echo because when I remove that cleanup step, all the echo commands work fine.  Going to call this resolved and dig into why cleanup is messing things up. Bizarre.</p>
<pre><code>      - name: Clean Workspace
        run: sudo rm -rf $GITHUB_WORKSPACE
</code></pre>

**Comment by rboekweg** [comment]
<p>After more testing, here are my findings.</p>
<ol>
<li>
<p>Self-hosted runners don’t start with clean workspace (<a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/about-self-hosted-runners" rel="nofollow noopener">see documentation</a>), so step to cleanup had been added</p>
</li>
<li>
<p>Any shell commands (e.g. echo set-env) that happen AFTER cleanup (sudo rm -rf $GITHUB_WORKSPACE) will fail with “No such file or directory”</p>
</li>
<li>
<p>If add checkout step before any shell commands, the echo set-env will work as expected.  There must be some additional context (setting path) that gets clobbered with rm command.</p>
</li>
</ol>

**Comment by hktv-tonyNg** [comment]
<p>thanks rboekweg for the solution that it is very strange that we need checkout code first (add actions/checkout@v2) to make setup env work…<br>
this waste me a day to test why cannot set env variable as i comment out all  checkout  code just try set variable syntax</p>
